lisp/ess-r-package.el

```diff
@@ -102,33 +101,32 @@ all source dirs recursively within the current package.")
 A project instance is a cons cell of the project type as symbol
 and the project path as string. If DIR is provided, the package
 is searched from that directory instead of `default-directory'."
-  (if (car ess-r-package--project-cache)
-      ess-r-package--project-cache
-    (let* ((pkg-path (ess-r-package--find-package-path (or dir default-directory)))
-           (project (when pkg-path
-                      (cons 'ess-r-package pkg-path))))
-      ;; Cache info for better performance on remotes
-      (setq-local ess-r-package--project-cache (or project (list nil)))
-      (when (car project)
-        project))))
+  (let ((pkg-info (ess-r-package-info dir)))
+    (when (car pkg-info)
+      (cons 'ess-r-package (cdr pkg-info)))))
 
 (cl-defmethod project-roots ((project (head ess-r-package)))
   "Return the project root for ESS R packages"
   (list (cdr project)))
 
 (defun ess-r-package-name (&optional dir)
   "Return the name of the current package as a string."
-  (when-let ((project (ess-r-package-project dir)))
-    (symbol-name (ess-r-package--find-package-name (cdr project)))))
-
-(defun ess-r-package-get-info ()
-  "Deprecated function to get package info.
-Please use `ess-r-package-project' instead."
-  (let ((project (ess-r-package-project)))
-    (if project
-        (cons (ess-r-package-name) (cdr project))
-      (list nil))))
-(make-obsolete 'ess-r-package-get-info 'ess-r-package-project "17.11")
+  (car (ess-r-package-info dir)))
+
+(defun ess-r-package-info (&optional dir)
+  "Get the description of the R project in directory DIR.
+Return a cons of the form (pkg-name . pkg-path). This value is
+cached for efficiency reasons."
+  (if (and (null dir) (car ess-r-package--project-cache))
+      ess-r-package--project-cache
+    (let* ((pkg-path (ess-r-package--find-package-path (or dir default-directory)))
+           (pkg-name (when pkg-path
+                       (ess-r-package--find-package-name pkg-path)))
+           (pkg-info (cons pkg-name pkg-path)))
+      ;; don't cache if DIR was supplied
```

Comment:
(Did we settle on sentence case for comments?)

Because `dir` might not be the buffer-local directory right? Perhaps slightly better would be to cache based on expanded directory, though that doesn't seem very important. Perhaps just make the comment clearer about the choices involved?

Reply:
Hash seems more trouble here because it would need to be cleared with some logic and deal somehow with stale paths. With buffer locals the user can revert the buffer and reset the cache.